src/log-store/src/raft_engine/log_store.rs

```diff
@@ -0,0 +1,519 @@
+        *self.cancel_token.lock().await = Some(token);
+        *self.gc_task_handle.lock().await = Some(handle);
+        self.started.store(true, Ordering::Relaxed);
+        Ok(())
+    }
+
+    async fn stop(&self) -> Result<(), Self::Error> {
+        let handle = self
+            .gc_task_handle
+            .lock()
+            .await
+            .take()
+            .context(IllegalStateSnafu)?;
+        let token = self
+            .cancel_token
+            .lock()
+            .await
+            .take()
+            .context(IllegalStateSnafu)?;
+        token.cancel();
+        handle.await.context(WaitGcTaskStopSnafu)?;
+        *self.engine.write().await = None;
+        self.started.store(false, Ordering::Relaxed);
+        info!("RaftEngineLogstore stopped");
+        Ok(())
+    }
+
+    /// Append an entry to logstore. Currently of existence of entry's namespace is not checked.
+    async fn append(&self, e: Self::Entry) -> Result<AppendResponse, Self::Error> {
+        let entry_id = e.id;
+        let mut batch = LogBatch::with_capacity(1);
+        batch
+            .add_entries::<MessageType>(e.namespace_id, &[e])
+            .context(AddEntryLogBatchSnafu)?;
+        let guard = self.engine.read().await;
+        let engine = guard.as_ref().context(IllegalStateSnafu)?;
+        engine.write(&mut batch, false).context(RaftEngineSnafu)?;
+        Ok(AppendResponse { entry_id })
+    }
+
+    /// Append a batch of entries to logstore. `RaftEngineLogstore` assures the atomicity of
+    /// batch append.
+    async fn append_batch(
+        &self,
+        ns: &Self::Namespace,
+        entries: Vec<Self::Entry>,
+    ) -> Result<Vec<Id>, Self::Error> {
+        let entry_ids = entries.iter().map(Entry::get_id).collect::<Vec<_>>();
+        let mut batch = LogBatch::with_capacity(entries.len());
+        batch
+            .add_entries::<MessageType>(ns.id, &entries)
+            .context(AddEntryLogBatchSnafu)?;
+        self.engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .write(&mut batch, false)
+            .context(RaftEngineSnafu)?;
+        Ok(entry_ids)
+    }
+
+    /// Create a stream of entries from logstore in the given namespace. The end of stream is
+    /// determined by the current "last index" of the namespace.
+    async fn read(
+        &self,
+        ns: &Self::Namespace,
+        id: Id,
+    ) -> Result<SendableEntryStream<'_, Self::Entry, Self::Error>, Self::Error> {
+        let engine = self
+            .engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .clone();
+        let last_index = engine.last_index(ns.id).unwrap();
+
+        let max_batch_size = self.config.read_batch_size;
+        let (tx, mut rx) = tokio::sync::mpsc::channel(max_batch_size);
+        let ns = ns.clone();
+        common_runtime::spawn_read(async move {
```

Comment:
Maybe we don't need to spawn now? Since we only read logs when opening a region, it should be fine to block the reader thread, I suppose.

Reply:
Logstore don't preassume where callers call `read`. We may also revisit the design of thread model as soon as we have done storage layer benchmark.